Dashboard Saturation Tests

rafabailon opened the pull request:
# Description

This PR includes the development done to implement `Dashboard Saturation Tests`, as well as the documentation related to its use.

Related to https://github.com/wazuh/wazuh/issues/24684

## Testing performed

```console
root@wazuh# dashboard-saturation-tests -p password -i 172.16.1.36
```

Result: [report.zip](https://github.com/user-attachments/files/16542340/report.zip)

rafabailon committed:
feat: add script, config files and tests

rafabailon committed:
feat: add README

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artillery.yml:
```diff
@@ -0,0 +1,37 @@
+config:
+  # Dashboard Base URL
+  target: https://127.0.0.1/
```
The target is currently hardcoded. To enhance the tool's versatility, dashboard tests should be able to target and saturate remote hosts.

rafabailon replied:
The target is always changed, it is a mandatory change in the Python script. I have removed the line from the YAML to avoid confusion. Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+import argparse
```
Please review the PEP 8 compliance and docstring formats for all Python files.
```
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:1:1: D100 Missing docstring in public module
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:1:1: I001 [*] Import block is un-sorted or un-formatted
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:14:5: D417 Missing argument description in the docstring for `format_directory`: `directory`
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:15:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:16:5: D410 [*] Missing blank line after section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:16:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:18:5: D411 [*] Missing blank line before section ("Returns")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:24:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:25:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:30:53: W291 [*] Trailing whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:31:1: W293 [*] Blank line contains whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:33:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:34:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:41:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:42:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:52:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:53:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:75:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:76:5: D410 [*] Missing blank line after section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:76:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:79:5: D411 [*] Missing blank line before section ("Returns")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:85:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:86:5: D410 [*] Missing blank line after section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:86:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:88:5: D411 [*] Missing blank line before section ("Returns")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:101:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:102:5: D410 [*] Missing blank line after section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:102:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:104:5: D411 [*] Missing blank line before section ("Returns")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:107:47: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:110:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:111:5: D410 [*] Missing blank line after section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:111:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:113:5: D411 [*] Missing blank line before section ("Returns")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:119:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:120:5: D410 [*] Missing blank line after section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:120:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:123:5: D411 [*] Missing blank line before section ("Returns")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:129:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:130:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:144:5: D205 1 blank line required between summary line and description
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:145:5: D411 [*] Missing blank line before section ("Args")
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:157:17: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:167:5: D103 Missing docstring in public function
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:168:65: W291 [*] Trailing whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:173:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:174:1: W293 [*] Blank line contains whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:176:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:177:1: W293 [*] Blank line contains whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:179:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:182:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:185:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:188:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:191:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:192:1: W293 [*] Blank line contains whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:195:1: W293 [*] Blank line contains whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:197:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:201:30: F541 [*] f-string without any placeholders
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:202:1: W293 [*] Blank line contains whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:208:5: D103 Missing docstring in public function
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:213:9: B007 Loop control variable `n` not used within loop body
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:213:45: W291 [*] Trailing whitespace
deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:217:11: W292 [*] No newline at end of file
 
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/4fee113a6384ad0a9d5dd204ecf37708f0632be5

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+csv_types = ['aggregate', 'intermediate']
+url_format = 'https://'
+
+def format_directory(directory):
+    """Check the paths of the directories and format them if necessary.
+    Args:
+        args (string): Path of a directory.
+    Returns:
+        string: Formatted path.
+    """
+    return directory if directory.endswith("/") else f'{directory}/'
```
Use `os.path.join(directory, '') `instead.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+        types (list): List of types.
+    """
+    for value in types:
+        if (value not in csv_types):
```
Brackets are not necessary

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+from subprocess import run
+
+# Global Configuration Variables
+csv_types = ['aggregate', 'intermediate']
```
This name is a little bit confussing. Please consider changing them or including descriptions for clarity.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+    """
+    return directory if directory.endswith("/") else f'{directory}/'
+
+def check_csv_types(types):
```
Please consider including this in the validation_scripts parameters functions

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+    return directory if directory.endswith("/") else f'{directory}/'
+
+def check_csv_types(types):
+    """Check that the CSV type is valid.
```
The term 'CSV type' is confusing. Please consider renaming it to make it clearer and easier to understand

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+    """
+    for value in types:
+        if (value not in csv_types):
+            raise Exception("CSV type is not valid") 
```
Consider including the invalid CSV type to the exception to improve debugging

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+        if (value not in csv_types):
+            raise Exception("CSV type is not valid") 
+        
+def create_directories(directory):
+    """Create directory that does not exist.
+    Args:
+        directory (str): Path of directory.
+    """
+    if not exists(directory):
+        makedirs(directory)
+
```
Use this instead `os.makedirs(directory, exist_ok=True)`

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+        makedirs(directory)
+
+def create_session_file(path, user):
+    """Create file to save the browser session.
+    Args:
+        path (str): Path of directory.
+        user (str): Username.
+    """
```
Consider improving docstring readability. Please apply similar improvement to all the docstrings 
```suggestion
   """Creates a file to save the browser session.

    Args:
        path (str): The directory path where the session file will be saved.
        user (str): The username associated with the session.

    Returns:
        None
    """
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/4fee113a6384ad0a9d5dd204ecf37708f0632be5

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+    check_csv_types(args.type)
+
+    # Create Directories
+    create_directories(args.logs)
+    create_directories(args.screenshots)
+    create_directories(args.csv)
+    create_directories(args.session)
```
Consider encapsulate initializaiton of directories inside it's own function

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+    """
+    return f"\"{key}\": \"{value}\""
+
+def gen_artillery_params(args):
+    """Format all parameters for Artillery.
+    Args:
+        args (ArgumentParser): Script parameters.
+    Returns:
+        str: Formatted Artillery parameters.
+    """
+    user = format_artillery_params('username', args.user)
+    password = format_artillery_params('password', args.password)
+    screenshots = format_artillery_params('screenshots', args.screenshots)
+    session = format_artillery_params('session', args.session)
+
+    params = f"\'{{{user}, {password}, {screenshots}, {session}}}\'"
+
+    return params
```
Consider launching artillery via a configuration file instead of passing parameters directly. This approach will help avoid error-prone logic

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+
+# Global Configuration Variables
+csv_types = ['aggregate', 'intermediate']
+url_format = 'https://'
```
Is this used exclusively in the `gen_url` function? If so, consider relocating it there.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,217 @@
+
+    process_script_arguments(script_args)
+
+    for n in range(0, script_args.quantity):    
+        run_artillery(script_args)
```
Consider adding a configurable timeout between iterations. Additionally, please rename the 'quantity' variable to something more descriptive, as its current name is unclear

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/ca020ab5fd076f3ff705605233b10f1ca0921806 and https://github.com/wazuh/wazuh-qa/pull/5629/commits/04ae05abe434b8be4ddae4b72fb582503ab38602

Rebits reviewed (changes requested):
Great job! This is very close to the initial requirements. There are a few areas where improvements could be made. Additionally, consider incorporating this library into a package to streamline automation.

rafabailon committed:
fix: refactor functions, add sleep and fix doc

rafabailon committed:
fix: change quantity variable

rafabailon committed:
fix: changes  according to the PEP 8 standard

rafabailon committed:
feat: add type hinting

rafabailon committed:
feat: add info to doc

rafabailon committed:
feat: improve gen_artillery_params function

rafabailon committed:
feat: add parameters to yaml automatically

rafabailon committed:
feat: add setup, requirements and organize code

rafabailon committed:
feat: add tests for alerts

rafabailon committed:
fix: change the event test and fix the documentation

rafabailon committed:
feat: change setup.py to pyproject.toml

rafabailon commented:
## Change `setup.py` to `pyproject.py`

I've changed the `setup.py` file to `pyproject.toml`. In the `.toml` file I've incorporated all the information needed to create and install the package (including Python dependencies).

The dependencies can be put in a `requirements.txt` file and referenced in `pyproject.toml`. I've chosen to include everything in the same file to make it easier to understand.

Below I leave the tests that I have done to verify that the `pyproject.toml` file works and has everything necessary to run the Python script.

<details>
   <summary>Code Execution</summary>
   <br/>

```console
root@ubuntu2204:/home/vagrant# python3 -m venv e2e
root@ubuntu2204:/home/vagrant# source e2e/bin/activate
(e2e) root@ubuntu2204:/home/vagrant# cd dashboard_saturation_tests/
(e2e) root@ubuntu2204:/home/vagrant/dashboard_saturation_tests# ls
data  pyproject.toml  README.md
(e2e) root@ubuntu2204:/home/vagrant/dashboard_saturation_tests# python3 -m pip install .
Processing /home/vagrant/dashboard_saturation_tests
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting pycodestyle>=2.12.1
  Downloading pycodestyle-2.12.1-py2.py3-none-any.whl (31 kB)
Collecting pandas>=2.2.2
  Downloading pandas-2.2.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.0/13.0 MB 34.9 MB/s eta 0:00:00
Collecting setuptools>=72.1.0
  Using cached setuptools-72.1.0-py3-none-any.whl (2.3 MB)
Collecting pyyaml>=6.0.2
  Downloading PyYAML-6.0.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (751 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 751.2/751.2 KB 37.3 MB/s eta 0:00:00
Collecting pytz>=2020.1
  Downloading pytz-2024.1-py2.py3-none-any.whl (505 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 505.5/505.5 KB 32.1 MB/s eta 0:00:00
Collecting tzdata>=2022.7
  Downloading tzdata-2024.1-py2.py3-none-any.whl (345 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.4/345.4 KB 47.6 MB/s eta 0:00:00
Collecting python-dateutil>=2.8.2
  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 229.9/229.9 KB 30.7 MB/s eta 0:00:00
Collecting numpy>=1.22.4
  Downloading numpy-2.0.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (19.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.5/19.5 MB 33.0 MB/s eta 0:00:00
Collecting six>=1.5
  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
Building wheels for collected packages: dashboard_saturation_tests
  Building wheel for dashboard_saturation_tests (pyproject.toml) ... done
  Created wheel for dashboard_saturation_tests: filename=dashboard_saturation_tests-1.0.0-py3-none-any.whl size=6093 sha256=8a381d54af825c5f1c073ef98b08aa7b7d738705d237b6d0ed275384f6c4e940
  Stored in directory: /tmp/pip-ephem-wheel-cache-aqt7te5d/wheels/10/f7/6a/88edb009f146e08bdcf6b4a1aa375bdf729b6c591c046a15fe
Successfully built dashboard_saturation_tests
Installing collected packages: pytz, tzdata, six, setuptools, pyyaml, pycodestyle, numpy, python-dateutil, pandas, dashboard_saturation_tests
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Uninstalling setuptools-59.6.0:
      Successfully uninstalled setuptools-59.6.0
Successfully installed dashboard_saturation_tests-1.0.0 numpy-2.0.1 pandas-2.2.2 pycodestyle-2.12.1 python-dateutil-2.9.0.post0 pytz-2024.1 pyyaml-6.0.2 setuptools-72.1.0 six-1.16.0 tzdata-2024.1
(e2e) root@ubuntu2204:/home/vagrant/dashboard_saturation_tests# dashboard-saturation-tests -p sMkJ*eFRgeWCizWUu8uENzcJzF?NmjGw -i 172.16.1.36
Test run id: tmpmb_6m5ayzfgzewcd7xjh399dd9jbryfw_agth
```

</details>

MARCOSD4 reviewed (changes requested):
Great job! Here are some suggestions to improve the code.

MARCOSD4 commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/README.md:
```diff
@@ -0,0 +1,114 @@
+│   ├── artillery.xml
+│   ├── dashboard_saturation_tests.py
+│   └── processor.js
+├── README.md
+├── requirements.txt
+├── setup.py
+└── version.json
```
It is necessary to update the directory structure as some have changed or are no longer there (`requirements.txt`, `setup.py`, etc.).

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/7b6d04e9738b6ecfa7b067089d145b2ee5dd4603

MARCOSD4 commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,367 @@
+        formatter_class=argparse.RawTextHelpFormatter
+    )
+
+    parser.add_argument(
+        '-l', '--log',
+        dest='logs',
+        default='logs/',
+        help=f'Directory to store the logs. Default "logs".'
+    )
```
I think you should include the type of parameter (`str`) since you have put it in some of them, it should appear in all of them.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/7b6d04e9738b6ecfa7b067089d145b2ee5dd4603

MARCOSD4 commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/tests/EndpointTest.js:
```diff
@@ -0,0 +1,71 @@
+class EndpointTest {
+
+    page = null;
+    vuContext = null;
```
What is `vuContext`? I think it's not entirely clear, consider adding documentation to this or using a more descriptive name.

rafabailon replied:
`vuContext` contains the information that is passed to the tests. All variables in `artillery.yml` are accessible using vuContext. As for `page` it is an object with the Browser information. The documentation is [here](https://playwright.dev/docs/api/class-page)

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/f46b2fc7da524084f1ae359a11bc2f0f6a02413c

MARCOSD4 commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/tests/EndpointTest.js:
```diff
@@ -0,0 +1,71 @@
+
+        await this.screenshotManager.takeAnScreenshot('test_03_endpoint_summary_is_loaded');
+
+        await expect(this.page.getByText('Endpoints')).toBeVisible();
```
I think that these functions should be commented explaining their purpose, as it is not entirely clear.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/f46b2fc7da524084f1ae359a11bc2f0f6a02413c

MARCOSD4 commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/tests/EventTest.js:
```diff
@@ -0,0 +1,77 @@
+
+        await this.screenshotManager.takeAnScreenshot('test_04_events_tab_is_loaded');
+
+        expect(await this.page.getByLabel('tr').count()).toBeGreaterThanOrEqual(1);
```
Same here.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/f46b2fc7da524084f1ae359a11bc2f0f6a02413c

MARCOSD4 commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/pyproject.toml:
```diff
@@ -0,0 +1,29 @@
+    "pycodestyle>=2.12.1",
+    "pyyaml>=6.0.2"
+]
+requires-python = ">=3.8"
```
```suggestion
requires-python = ">=3.10"
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/7b6d04e9738b6ecfa7b067089d145b2ee5dd4603

MARCOSD4 commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,367 @@
+import argparse
```
You must include the copyright section and a docstring explaining the function of the module, following the PEP8 standards.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/7b6d04e9738b6ecfa7b067089d145b2ee5dd4603

MARCOSD4 replied:
The copyright must be the same that is being used in the rest of Wazuh's projects, to follow the same standard in all of them. You can see this [example file](https://github.com/wazuh/wazuh-qa/blob/9eba4fb752e2d4b8862e9a96c7db9c0afbc8d9c5/tests/end_to_end/test_vulnerability_detector/test_vulnerability_detector.py#L1-L3).

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/54c17e97c71cb3501f7eb006beeb9f8692b8bbf7

rafabailon commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/README.md:
```diff
@@ -0,0 +1,114 @@
+
+- `--logs`, `--screenshots`, `--csv`, and `--session` are directories (if they do not exist, they are created in the Script).
+- `--artillery` needs to receive a valid Artillery configuration file.
+- `--type` only accepts the options received in the Artillery JSON file with the results.
```
The Artillery configuration file is `artillery.yml`. It can be called whatever you like since it is passed to Artillery via a command. I have named it `artillery.yml` but the configuration file can be called whatever you like and even have multiple files.

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/7b6d04e9738b6ecfa7b067089d145b2ee5dd4603

rafabailon committed:
fix: update the python version, add the coypwright and fix the readme

rafabailon committed:
feat: add documentation in JSDoc format

MARCOSD4 reviewed (changes requested):
Minor change

rafabailon committed:
feat: add copyright to the beginning of the script

rafabailon committed:
feat: remove unused coypwright

MARCOSD4 reviewed (approved):
LGTM

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+from datetime import datetime
+from os import makedirs, getcwd
+from os.path import join, isabs
+from subprocess import run
```
```suggestion
import json
import time
from argparse import Namespace
from datetime import datetime
from os import getcwd, makedirs
from os.path import isabs, join
from subprocess import run

import pandas as pd
import yaml
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits replied:
Remove the import of argparse
```python
import argparse
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/bf1f8f0173f4aa3021af6bba3714978f4716802b

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+
+def check_artillery_result_types(types: list) -> None:
+    """Check that a valid type of Artillery result has been
+    chosen to generate the CSV.
```
Follow https://google.github.io/styleguide/pyguide.html#382-modules. One line for the summary followed by a white line before a full description if needed

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+    """
+    for value in types:
+        if value not in artillery_result_type:
+            msg = f'The Artillery result type to generate the CSV is not ' \
```
```suggestion
            msg = 'The Artillery result type to generate the CSV is not ' \
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+    """Check the paths of the directories and format them if necessary.
+
+    Args:
+        args (str): Path of a directory.
```
```suggestion
        directory (str): Path of a directory.
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+    makedirs(directory, exist_ok=True)
+
+
+def provide_directories(args: argparse) -> None:
```
argparse is the module's name. Here we need to use Namespace from argparse module

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+
+
+def provide_directories(args: argparse) -> None:
+    """Format the paths of the directories and create them
```
Follow https://google.github.io/styleguide/pyguide.html#382-modules. One line for the summary followed by a white line before a full description if needed

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+        file.write('{}')
+
+
+def gen_artillery_params(args: argparse) -> dict:
```
Same here, use Namespace

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+        yaml.dump(data, f)
+
+
+def process_script_arguments(args: argparse) -> None:
```
Same here, use Namespace

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+    Returns:
+        str: File name of the log (include path).
+    """
+    return log_path + datetime.now().strftime(f"log-%Y%m%d%H%M%S.log")
```
```suggestion
    return log_path + datetime.now().strftime("log-%Y%m%d%H%M%S.log")
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+    return csv_path + datetime.now().strftime(f"{type}-%Y%m%d%H%M%S.csv")
+
+
+def convert_json_to_csv(args: argparse, json_output: str) -> None:
```
Same here. Use Namespace

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+        df.to_csv(csv_filename, index=False)
+
+
+def run_artillery(args: argparse) -> None:
```
Same here, use Namespace

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+    convert_json_to_csv(args, json_filename)
+
+
+def get_script_arguments() -> argparse:
```
Same here. Use Namespace

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+
+
+def main() -> None:
+    """Run the Script.
+    """
```
```suggestion
    """Run the Script."""
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,382 @@
+
+    process_script_arguments(script_args)
+
+    for n in range(0, script_args.iterations):
+        run_artillery(script_args)
+        time.sleep(script_args.wait)
+
```
```suggestion
    for _ in range(0, script_args.iterations):
        run_artillery(script_args)
        time.sleep(script_args.wait)

```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/a07ebb3ed73c7bc3ab8fce58bf98032b832575f8

Rebits reviewed (changes requested):
## Usage and documentation
 
These tests are planned to be launched in a node in the CI process. Differencing other tests, these will require an environment with a GUI along with other dependencies
However, this is not commented on in the documentation and there is no easy way to deploy and provision an environment where these tests can be launched.

I couldn't launch these tests on either my local or a virtual host due to the following errors:

<details>

<summary> Local </summary>

Playwright is not supported in the OS, leading into unexpected error

```shell
N: This must be accepted explicitly before updates for this repository can be applied. See apt-secure(8) manpage for details.
Failed to install browsers
Error: Installation process exited with code: 100
```


</details>


<details>

<summary> Ubuntu Virtual Environment </summary>

```

```
Test run id: tczfj_8n8qq8a3bcw7hgchnwtq4h8zbazfb_bfam
⠋ node:internal/modules/cjs/loader:1225
  const err = new Error(message);
              ^

Error: Cannot find module '@playwright/test'
Require stack:
- /home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/tests/LoginTest.js
- /home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/processor.js
- /usr/lib/node_modules/artillery/node_modules/@artilleryio/int-core/lib/runner.js
- /usr/lib/node_modules/artillery/node_modules/@artilleryio/int-core/index.js
- /usr/lib/node_modules/artillery/lib/artillery-global.js
- /usr/lib/node_modules/artillery/bin/run
    at Module._resolveFilename (node:internal/modules/cjs/loader:1225:15)
    at Module._load (node:internal/modules/cjs/loader:1051:27)
    at Module.require (node:internal/modules/cjs/loader:1311:19)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/tests/LoginTest.js:1:20)
    at Module._compile (node:internal/modules/cjs/loader:1469:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
    at Module._load (node:internal/modules/cjs/loader:1104:12)
    at Module.require (node:internal/modules/cjs/loader:1311:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/tests/LoginTest.js',
    '/home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/processor.js',
    '/usr/lib/node_modules/artillery/node_modules/@artilleryio/int-core/lib/runner.js',
    '/usr/lib/node_modules/artillery/node_modules/@artilleryio/int-core/index.js',
    '/usr/lib/node_modules/artillery/lib/artillery-global.js',
    '/usr/lib/node_modules/artillery/bin/run'
  ]
}

Node.js v20.17.0
Traceback (most recent call last):
  File "/home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/./data/dashboard_saturation_tests.py", line 382, in <module>
    main()
  File "/home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/./data/dashboard_saturation_tests.py", line 377, in main
    run_artillery(script_args)
  File "/home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/./data/dashboard_saturation_tests.py", line 253, in run_artillery
    convert_json_to_csv(args, json_filename)
  File "/home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/./data/dashboard_saturation_tests.py", line 221, in convert_json_to_csv
    with open(json_output) as f:
FileNotFoundError: [Errno 2] No such file or directory: 'logs/log-20240828105032.log'
(example-final) vagrant@ubuntu22:~/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests$ npm install @playwright/test

added 3 packages in 4s
```

```

</details>

Given the complexity of the dependencies and the environment-specific requirements, we should consider creating a provisioning script or even a Docker image. This would allow us to run the tests with specified parameters and generate the reports in CSV format. Implementing this solution will streamline future integrations. Otherwise, managing dependencies and setting up specific nodes with a GUI will be challenging.

## Functionality

It was not possible to test the development. This will be reviewed when the usage has been refactored in order to be launched in any environment

## Style

Some minor style changes have been requested

## Packaging 

The packages should only include the necessary dependencies and libraries, not the tests themselves. Additionally, I encountered some issues during the installation of the package.
```shell
vagrant@ubuntu22:~/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests$ source /tmp/example-final/bin/activate
(example-final) vagrant@ubuntu22:~/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests$ python3 -m pip install .
Processing /home/vagrant/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      error: Multiple top-level packages discovered in a flat-layout: ['csv', 'logs', 'data', 'screenshots'].
      
      To avoid accidental inclusion of unwanted files or directories,
      setuptools will not proceed with this build.
      
      If you are trying to create a single distribution with multiple packages
      on purpose, you should not rely on automatic discovery.
      Instead, consider the following options:
      
      1. set up custom discovery (`find` directive with `include` or `exclude`)
      2. use a `src-layout`
      3. explicitly set `py_modules` or `packages` with a list of names
      
      To find more information, look for "package discovery" on setuptools docs.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

rafabailon committed:
fix: suggested changes and fixes

rafabailon committed:
feat: add docker, modify pyproject and documentation

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,388 @@
+# This program is a free software; you can redistribute it and/or
+# modify it under the terms of GPLv2
+
+"""Dashboard Saturation Tests Module
```
```suggestion
"""Dashboard Saturation Tests Module.
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/facce46e13cad8a5998ec51bb95e70076e637267

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,388 @@
+    quiet = ''
+
+    if not args.debug:
+        quiet = f'-q'
```
```suggestion
        quiet = '-q'
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/d3d4bf232f928d98091163ba2aaec83ca14a5cae

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,388 @@
+from subprocess import run
+
+import pandas as pd
+import yaml
```
Impots shoud follow pep8: https://peps.python.org/pep-0008/#imports

> Imports should be grouped in the following order:
> 
>     Standard library imports.
>     Related third party imports.
>     Local application/library specific imports.
> 
> You should put a blank line between each group of imports.

No extra white line is necessary between time and argpase imports.
In addition remove the import of the Namespace importing the argparse directly or import along with the Namespace the ArgumentParsers:
```python
import argparse
```

```suggestion
import json
import time
from argparse import Namespace
from datetime import datetime
from os import getcwd, makedirs
from os.path import isabs, join
from subprocess import run

import pandas as pd
import yaml
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/56f5d2f0bf8080dbe3bf7f1f58d5c7b3420d84f3

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,388 @@
+    Returns:
+        argparse: Script parameters.
+    """
+
```
Fulfill with PEP257 for docstrings https://peps.python.org/pep-0257/
> No blank lines allowed after function docstring

```suggestion
```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/facce46e13cad8a5998ec51bb95e70076e637267

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,388 @@
+        default=5,
+        help=f'Waiting Time between Executions.'
+    )
+
```
f-string without any placeholders

```suggestion
    parser.add_argument(
        '-l', '--log',
        dest='logs',
        type=str,
        default='logs/',
        help='Directory to store the logs. Default "logs".'
    )

    parser.add_argument(
        '-s', '--screenshots',
        dest='screenshots',
        type=str,
        default='screenshots/',
        help='Directory to store the screenshots. Default "screenshots".')

    parser.add_argument(
        '-c', '--csv',
        dest='csv',
        type=str,
        default='csv/',
        help='Directory to store the CSVs. Default "csv".'
    )

    parser.add_argument(
        '-o', '--session',
        dest='session',
        type=str,
        default='.auth/',
        help='Directory to store the Sessions. Default ".auth".'
    )

    parser.add_argument(
        '-a', '--artillery',
        dest='artillery',
        type=str,
        default="data/artillery.yml",
        help='Path to the Artillery Script. Default "artillery.yml".'
    )

    parser.add_argument(
        '-u', '--user',
        dest='user',
        type=str,
        default='admin',
        help='Wazuh User for the Dashboard. Default "admin".'
    )

    parser.add_argument(
        '-p', '--password',
        dest='password',
        type=str,
        required=True,
        help='Wazuh Password for the Dashboard.'
    )

    parser.add_argument(
        '-q', '--iterations',
        dest='iterations',
        type=int,
        default=1,
        help='Number of Tests to Run. Default {1}'
    )

    parser.add_argument(
        '-i', '--ip',
        dest='ip',
        type=str,
        required=True,
        help='Set the Dashboard IP.'
    )

    parser.add_argument(
        '-t', '--type',
        dest='type',
        type=str,
        nargs='+',
        action='store',
        default=['aggregate', 'intermediate'],
        help='JSON data to create the CSV.'
    )

    parser.add_argument(
        '-w', '--wait',
        dest='wait',
        type=int,
        default=5,
        help='Waiting Time between Executions.'
    )

```

rafabailon replied:
Solved in https://github.com/wazuh/wazuh-qa/pull/5629/commits/e25de36fff40092faa79587bacea3e7cd1c8aa2e

rafabailon committed:
fix: remove argparse import

Rebits reviewed (changes requested):
Docker image, makefiles, or package construction seems to not work as expected. Please check it
```shell
(dashboard) ➜  dashboard_saturation_tests git:(enhancement/24684-dashboard-saturation-tests) make test COMMAND_OPTIONS='--password wT2duXovSMCKiRi348yb72lhmRQ8wI*1 --ip 192.168.56.8'

docker exec -it dashboard_saturation_tests /bin/bash -c "dashboard-saturation-tests --password wT2duXovSMCKiRi348yb72lhmRQ8wI*1 --ip 192.168.56.8" || true
Traceback (most recent call last):
  File "/usr/local/bin/dashboard-saturation-tests", line 5, in <module>
    from data.dashboard_saturation_tests import main
ModuleNotFoundError: No module named 'data'
mkdir -p artifacts
docker cp dashboard_saturation_tests:/app/dashboard_saturation_tests/logs /home/rebits/Wazuh/wazuh-qa/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artifacts
Error response from daemon: Could not find the file /app/dashboard_saturation_tests/logs in container dashboard_saturation_tests
make: *** [Makefile:35: artifacts] Error 1
```

rafabailon committed:
fix: fix documentation and adjust it to PEP257

rafabailon committed:
fix: fix get_script_arguments style and remove f-string

rafabailon committed:
fix: reorganize imports at the beginning of the document

rafabailon committed:
fix: remove f-string in run_artillery function

rafabailon committed:
fix: fix package error in pyproject and improve documentation

rafabailon commented:
@Rebits  I have fixed the comments you left me. I have also checked the problem with Docker. I have been able to reproduce it and the problem was in the packages. I have removed some unnecessary lines in `pyproject.toml` about the organization of the packages and I have tested again. Now it works correctly. I have also added more documentation so that it is well understood how everything works.

The commit with these changes is https://github.com/wazuh/wazuh-qa/pull/5629/commits/4949dbd60a4e62224ffcd539bee42577d043653c

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/README.md:
````diff
@@ -0,0 +1,131 @@
+- `--artillery` needs to receive a valid Artillery configuration file (for example, `artillery.yml`).
+- `--type` only accepts two values ​​(`aggregate` or `intermediate`). Either or both can be chosen.
+
+### Check PEP 8
+
+The Python script complies with the PEP 8 standard. To verify that it continues to comply with the standard (after making changes) you just have to execute the following commands:
+
+```shell script
+pycodestyle dashboard_saturation_tests.py
+```
````
Not needed. This should be part of the repository standard
```suggestion
```

Rebits replied:
The research and the improvement of these tests will be carried on in another issue.

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/README.md:
```diff
@@ -0,0 +1,131 @@
+# Dashboard Saturation Tests
+
+This module contains a basic set of tests for the Wazuh Dashboard. The tests are performed using Artillery and Playwright. The tests are executed through a Python script.
+
+The Python script needs some input parameters to perform the tests and get the results. Artillery gets the test results in JSON format and the script converts them to CSV format so they can be analyzed more easily.
+
```
It is unclear what specific tests are being conducted. Please specify that these are **performance** tests. Additionally, include a detailed section specifying exactly which views, endpoints, or components are being tested

Rebits replied:
The research and the improvement of these tests will be carried on in another issue.

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/README.md:
```diff
@@ -0,0 +1,131 @@
+
+### Install Wazuh
+
+You need to have a Wazuh installation (including the Dashboard). You also need (for some tests) Agents connected to the Manager. The Dashboard must be accessible and have the user data and password to log in.
+
+To run the script you need to have Python and Pip installed.
```
For which tests is it necessary to have the agent and why?

Rebits replied:
The research and the improvement of these tests will be carried on in another issue.

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/README.md:
```diff
@@ -0,0 +1,131 @@
+
+- `--logs`, `--screenshots`, `--csv`, and `--session` are directories (if they do not exist, they are created in the Script).
+- `--artillery` needs to receive a valid Artillery configuration file (for example, `artillery.yml`).
+- `--type` only accepts two values ​​(`aggregate` or `intermediate`). Either or both can be chosen.
```
What do aggregate and intermediate mean?

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/Makefile:
```diff
@@ -0,0 +1,56 @@
+	docker system prune -a -f
+.PHONY: destroy
+destroy: rm purge ## Destroy All Docker Resources
+.PHONY: style
+style: ## Check Python Style
+	docker exec -it $(APP) $(CONSOLE) -c "pycodestyle data/dashboard_saturation_tests.py" || true
```
Not necessary
```suggestion
```

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,387 @@
+
+# Global Configuration Variables
+artillery_result_type = [
+    # Final statistics printed to the console at the end of a test
+    'aggregate',
+    # Statistics that print to the console while a test is running
+    'intermediate'
```
Confusing description for the normal user. In addition this is not included in the README

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/README.md:
````diff
@@ -0,0 +1,131 @@
+dashboard-saturation-tests -p password -i ip
+```
+
+- Result: [report.zip](https://github.com/user-attachments/files/16542340/report.zip)
````
There is no explanation regarding the CSV structure and the data. This would be make it challenging to understand the results of the tests and to determinate the real performance results in the environment

Rebits replied:
The research and the improvement of these tests will be carried on in another issue.

Rebits commented on deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/data/dashboard_saturation_tests.py:
```diff
@@ -0,0 +1,387 @@
+    makedirs(directory, exist_ok=True)
+
+
+def provide_directories(args: Namespace) -> None:
```
```suggestion
def create_test_directories(args: Namespace) -> None:
```

Rebits reviewed (changes requested):
Good job, overall, the development provides a strong foundation for starting performance testing on the Dashboard component. However, there are a few issues that should be addressed before incorporating this into the performance testing process. These should be change in the specified issues:
  
# Documentation

- **Lack of hardware requirements in the documentation**

During local testing, all tests failed due to timeouts, likely caused by resource limitations in the environment. However, the README doesn’t provide any information regarding the required resources. How much resources are necessary to successfully run these tests? This is a critical question that needs to be addressed to automate these tests in the performance testing environment.

```shell
➜  dashboard_saturation_tests git:(enhancement/24684-dashboard-saturation-tests) make test COMMAND_OPTIONS='--password PASSWORD --ip IP'
docker exec -it dashboard_saturation_tests /bin/bash -c "dashboard-saturation-tests --password U9vwwe3D9ByRyL0Qzt*TPU?RP*CU*yGZ --ip 192.168.56.8" || true
Test run id: tbae4_388kdcdf3fxh56r88e3989rpgf7h3_8w3p 
⠦ page.waitForResponse: Timeout 30000ms exceeded while waiting for event "response"
    at PathManager.waitfor (/app/dashboard_saturation_tests/data/lib/PathManager.js:96:21)
    at async LoginTest.checkLogin (/app/dashboard_saturation_tests/data/tests/LoginTest.js:72:9)
    at async LoginTest.executeTest (/app/dashboard_saturation_tests/data/tests/LoginTest.js:89:9)
    at async test_login (/app/dashboard_saturation_tests/data/processor.js:12:5)
    at async Array.scenario (/usr/lib/node_modules/artillery/node_modules/artillery-engine-playwright/index.js:372:11) {
  name: 'TimeoutError'
}
⠙ page.waitForResponse: Timeout 30000ms exceeded while waiting for event "response"
    at PathManager.waitfor (/app/dashboard_saturation_tests/data/lib/PathManager.js:96:21)
    at async LoginTest.checkLogin (/app/dashboard_saturation_tests/data/tests/LoginTest.js:72:9)
    at async LoginTest.executeTest (/app/dashboard_saturation_tests/data/tests/LoginTest.js:89:9)
    at async test_login (/app/dashboard_saturation_tests/data/processor.js:12:5)
    at async Array.scenario (/usr/lib/node_modules/artillery/node_modules/artillery-engine-playwright/index.js:372:11) {
  name: 'TimeoutError'
}
⠦ page.waitForResponse: Timeout 30000ms exceeded while waiting for event "response"
    at PathManager.waitfor (/app/dashboard_saturation_tests/data/lib/PathManager.js:96:21)
    at async EventTest.accessAgent (/app/dashboard_saturation_tests/data/tests/EventTest.js:64:9)
    at async EventTest.executeTest (/app/dashboard_saturation_tests/data/tests/EventTest.js:106:9)
    at async test_accessEvent (/app/dashboard_saturation_tests/data/processor.js:39:5)
    at async Array.scenario (/usr/lib/node_modules/artillery/node_modules/artillery-engine-playwright/index.js:372:11) {
  name: 'TimeoutError'
}
⠧ page.waitForResponse: Timeout 30000ms exceeded while waiting for event "response"
    at PathManager.waitfor (/app/dashboard_saturation_tests/data/lib/PathManager.js:96:21)
    at async EventTest.accessAgent (/app/dashboard_saturation_tests/data/tests/EventTest.js:64:9)
    at async EventTest.executeTest (/app/dashboard_saturation_tests/data/tests/EventTest.js:106:9)
    at async test_accessEvent (/app/dashboard_saturation_tests/data/processor.js:39:5)
    at async Array.scenario (/usr/lib/node_modules/artillery/node_modules/artillery-engine-playwright/index.js:372:11) {
  name: 'TimeoutError'
}
⠇ page.waitForResponse: Timeout 30000ms exceeded while waiting for event "response"
    at PathManager.waitfor (/app/dashboard_saturation_tests/data/lib/PathManager.js:96:21)
    at async EventTest.accessAgent (/app/dashboard_saturation_tests/data/tests/EventTest.js:64:9)
    at async EventTest.executeTest (/app/dashboard_saturation_tests/data/tests/EventTest.js:106:9)
    at async test_accessEvent (/app/dashboard_saturation_tests/data/processor.js:39:5)
    at async Array.scenario (/usr/lib/node_modules/artillery/node_modules/artillery-engine-playwright/index.js:372:11) {
  name: 'TimeoutError'
}
mkdir -p artifacts
docker cp dashboard_saturation_tests:/app/dashboard_saturation_tests/logs /home/rebits/Wazuh/wazuh-qa-final/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artifacts
Successfully copied 91.1kB to /home/rebits/Wazuh/wazuh-qa-final/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artifacts
docker cp dashboard_saturation_tests:/app/dashboard_saturation_tests/screenshots /home/rebits/Wazuh/wazuh-qa-final/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artifacts
Successfully copied 994kB to /home/rebits/Wazuh/wazuh-qa-final/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artifacts
docker cp dashboard_saturation_tests:/app/dashboard_saturation_tests/csv /home/rebits/Wazuh/wazuh-qa-final/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artifacts
Successfully copied 161kB to /home/rebits/Wazuh/wazuh-qa-final/deps/wazuh_testing/wazuh_testing/dashboard_saturation_tests/artifacts

```

- **Lack of test details**: There is no documentation outlining which specific tests are being executed, nor any information on which menus or dashboard components are being tested.

- **Lack of Artifact Output Documentation**: The provided output generates a large volume of data. It is essential to document how users can extract meaningful information from the results.


The research and the improvement of these tests will be carried on in another issue.

# Usage

## Report

- **Screenshots overwritten with each test iteration**: While logs and CSV reports are generated with unique timestamps, screenshots are being overwritten with each test iteration. I recommend implementing unique timestamps for screenshots as well and adjusting the timestamp format for better readability.
    
- **No Post-Test data processing**: After test completion, two CSV files are generated containing aggregate and intermediate metrics. While this data is necessary, no further processing is performed. The CSV files include raw data from Artillery, but we should extract key metrics and values to generate meaningful insights using the data analysis module.
    
- **CSV Results Incompatible with Data Analysis Module**: The current format of the CSV results is not compatible with the data analysis module and needs to be adjusted.
    
- **Logs and CSV contain redundant information**: The CSV reports should be structured differently from the raw data output by Artillery, as currently, both logs and CSV files contain the same information. Additionally, the Artillery logs should be renamed with a `.json` extension instead of `.log`.
    
- **Aggregate and Intermediate Metrics**: The reports generate two CSV files—one for aggregate data and one for intermediate data. Since aggregate data is a summary of the intermediate data, and intermediate metrics are primarily used for real-time monitoring, we should evaluate whether the final CSV report needs to include intermediate data or just the aggregated results.
    

All these considerations will be discussed and addressed in another issue.


# Style

Some missing documentation and minor style errors:
```
__init__.py:1:1: D104 Missing docstring in public package
data/__init__.py:1:1: D104 Missing docstring in public package
data/dashboard_saturation_tests.py:334:14: F541 [*] f-string without any placeholders
```

This should be carried on during the migration to the another repository
